libraries/chain/fork_database.cpp

```diff
@@ -603,15 +603,24 @@ namespace eosio::chain {
    }
 
    template<class BSP>
-   bool fork_database_t<BSP>::validated_block_exists(const block_id_type& id) const {
+   bool fork_database_t<BSP>::validated_block_exists(const block_id_type& id, uint32_t block_num) const {
       std::lock_guard g( my->mtx );
-      return my->validated_block_exists_impl(id);
+      return my->validated_block_exists_impl(id, block_num);
    }
 
+   // returns true if block `id`, or one of its ancestors >= block_num, is found in fork_db and `is_valid()`
+   // precondition: `id` is already in fork_db, so is a descendent of `root`
+   // ------------------------------------------------------------------------------------------------------
    template<class BSP>
-   bool fork_database_impl<BSP>::validated_block_exists_impl(const block_id_type& id) const {
-      auto itr = index.find( id );
-      return itr != index.end() && (*itr)->is_valid();
+   bool fork_database_impl<BSP>::validated_block_exists_impl(const block_id_type& id, uint32_t block_num) const {
+      assert(root && (root->id() == id || index.find(id) != index.end()));
+      if (root->block_num() >= block_num)
+         return true;
```

Comment:
I don't see how you can assume this is true.

If the block you are considering voting on was not on the best branch then this can return true when that branch has been abandoned and pruned from the forkdb.

I was expecting this to walk back from `id` until it hit root or a validated block or block_num < `block_header::num_from_id(latest_qc_claim__block_ref.block_id)` and verifying it reached `latest_qc_claim__block_ref.block_id` or root.

Reply:
> If the block you are considering voting on was not on the best branch then this can return true when that branch has been abandoned and pruned from the forkdb.

Yes, good point, I updated the PR.

Reply:
I think it is "correct" now, but seems overly convoluted and error prone. Seems to me much safer to pass in `latest_qc_claim__block_ref.block_id` which provides a range of blocks similar to `fetch_branch` that takes two ids. Also would be more efficient as you don't have to do all the `shared_ptr` copies.

Also. please add some unit tests for this new forkdb method.

Reply:
Personally, I find the `fetch_branch` that takes two ids more convoluted than the `fetch_branch` that takes an `id` and a `block_num`.
But I'll change `validated_block_exists_impl` to take two `id`s. 
I left the `assert` because it makes the code simpler and faster (allows to `return true;` at the end without checks).

Reply:
I think once you add some unit tests you will see the current implementation also allows for an abandoned branch to return true.

Reply:
Since `id` *is* in fork_db (per the assert), and walking back from `is` we don't find 'claimed_id`, it means that `claimed_id` is either root or and ancester of root (since we know that `claimed_id` is an ancestor of `id).

Reply:
```
b1 <- b2  <- b3  <-b4
      b2' <- b3'
```
* net-thread
 -- call `create_block_state_i`
  --- find bottom b3', create block state b4'
 * main-thread
-- advance root to top b2, purges b2', b3'
 net-thread
 -- continues `create_block_state_i` execution 
  --- call `consider_voting`
  --- calls `fork_db_validated_block_exists(b3', b1')
  --- returns `true`
  --- Also returns `true` if b2' or b3' happen to be valid.

Reply:
Thanks for explaining, I finally understand that the `assert` fay be false and `id` may not be present. https://github.com/AntelopeIO/spring/pull/788/commits/3a133d244e24c14f55b3b5ebdf4d936cabaf3b91